connectors/sources/gmail.py

```diff
@@ -0,0 +1,352 @@
+    }
+
+    SCHEMA = fastjsonschema.compile(
+        definition=SCHEMA_DEFINITION,
+    )
+
+    async def validate(self, advanced_rules):
+        if len(advanced_rules) == 0:
+            return SyncRuleValidationResult.valid_result(
+                SyncRuleValidationResult.ADVANCED_RULES
+            )
+
+        try:
+            GMailAdvancedRulesValidator.SCHEMA(advanced_rules)
+
+            return SyncRuleValidationResult.valid_result(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES
+            )
+        except JsonSchemaValueException as e:
+            return SyncRuleValidationResult(
+                rule_id=SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=e.message,
+            )
+
+
+def _message_doc(message):
+    timestamp_field = "_timestamp"
+
+    # We're using the `_attachment` field here so the attachment processor on the ES side decodes the base64 value
+    message_fields_to_es_doc_mappings = {
+        MessageFields.ID: "_id",
+        MessageFields.FULL_MESSAGE: "_attachment",
+        MessageFields.CREATION_DATE: timestamp_field,
+    }
+
+    es_doc = {
+        es_doc_field: message.get(message_field.value)
+        for message_field, es_doc_field in message_fields_to_es_doc_mappings.items()
+    }
+
+    # The attachment processor cannot handle base64url encoded values (only ordinary base64)
+    es_doc["_attachment"] = base64url_to_base64(es_doc["_attachment"])
+
+    if es_doc.get(timestamp_field) is None:
+        es_doc[timestamp_field] = iso_utc()
+
+    return es_doc
+
+
+class GMailDataSource(BaseDataSource):
+    """GMail"""
+
+    name = "GMail"
+    service_type = "gmail"
+    advanced_rules_enabled = True
+    dls_enabled = True
+
+    def __init__(self, configuration):
+        super().__init__(configuration=configuration)
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for the GMail connector.
+        Returns:
+            dict: Default configuration.
+        """
+        default_credentials = {
+            "type": "service_account",
+            "project_id": "dummy_project_id",
+            "private_key_id": "abc",
+            "private_key": "",
+            "client_email": "123-abc@developer.gserviceaccount.com",
+            "client_id": "123-abc.apps.googleusercontent.com",
+            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+            "token_uri": "http://localhost:4444/token",
+        }
+
+        return {
+            "service_account_credentials": {
+                "display": "textarea",
+                "label": "GMail service account JSON",
+                "order": 1,
+                "type": "str",
+                "value": json.dumps(default_credentials),
+                "required": True,
+            },
+            "subject": {
+                "display": "text",
+                "label": "Subject",
```

Comment:
Why this is called `Subject`? It's hard to get the meaning of `Admin account email address` from `Subject`.

Reply:
It's also called subject in the json credentials object provided by GMail. Therefore I've kept the same name for consistency.

Reply:
As the `label` is what we are going to display in Kibana UI, I think it makes sense to change it to something easier to understand.